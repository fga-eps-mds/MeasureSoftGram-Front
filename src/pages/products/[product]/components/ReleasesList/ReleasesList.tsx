import React from 'react';

import { Box, Typography, Container, Button } from '@mui/material';

import { productQuery } from '@services/product';
import { useProductContext } from '@contexts/ProductProvider';
import { useOrganizationContext } from '@contexts/OrganizationProvider';
import { useRequest } from '@hooks/useRequest';
import { CompareGoalAccomplished } from '@customTypes/product';
import { useRouter } from 'next/router';
import ReleasesTable from './ReleasesTable';

import * as Styles from './styles';
import Skeleton from './Skeleton';

function ReleasesList() {
  const { currentProduct } = useProductContext();
  const { currentOrganization } = useOrganizationContext();
  const router = useRouter();

  const { data: releaseList, isLoading } = useRequest<CompareGoalAccomplished[]>(
    productQuery.getReleaseList(currentOrganization?.id as string, currentProduct?.id as string)
  );
  if (isLoading) return <Skeleton />;

  function pushToReleasesPath() {
    const releasesPath = `/products/${currentOrganization?.id}-${currentProduct?.id}-${currentProduct?.name}/releases`;
    void router.push(releasesPath);
  }

  return (
    <Box display="flex" flexDirection="column" mt="42px">
      <Styles.Wrapper>
        <Container>
          <Typography variant="h5" marginRight="10px" mb="32px" color="#538BA3" fontWeight="500">
            Ãšltima Release criada
          </Typography>

          <ReleasesTable releaseList={releaseList?.slice(0, 1) ?? []} />
          <Box display="flex" flexDirection="column" mt="10px" alignItems="center">
            <Button onClick={() => void pushToReleasesPath()} variant="text">
              VER MAIS...
            </Button>
          </Box>
        </Container>
      </Styles.Wrapper>
    </Box>
  );
}

export default ReleasesList;
