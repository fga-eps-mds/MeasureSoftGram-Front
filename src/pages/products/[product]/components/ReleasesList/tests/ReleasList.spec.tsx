import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { useRequest } from '@hooks/useRequest';
import { useRouter } from 'next/router';
import ReleasesList from '../ReleasesList';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn()
  })
}));

jest.mock('@contexts/ProductProvider', () => ({
  useProductContext: jest.fn(() => ({
    currentProduct: {
      id: '1',
      name: 'MeasureSoftGram'
    }
  }))
}));

jest.mock('@contexts/OrganizationProvider', () => ({
  useOrganizationContext: jest.fn(() => ({
    currentOrganization: {
      id: '1'
    }
  }))
}));

jest.mock('@hooks/useRequest', () => ({
  useRequest: jest.fn(() => ({
    data: [],
    isLoading: false
  }))
}));

afterEach(cleanup);

describe('<ReleasesList />', () => {
  it('renders without crashing', () => {
    const { asFragment } = render(<ReleasesList />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the releases table component when not loading', () => {
    (useRequest as jest.Mock).mockReturnValueOnce({
      data: [
        { id: '1', name: 'Release 1' },
        { id: '2', name: 'Release 2' }
      ],
      isLoading: false
    });

    const { queryByText } = render(<ReleasesList />);

    expect(queryByText('Nome')).toBeInTheDocument();
    expect(queryByText('InÃ­cio da release')).toBeInTheDocument();
    expect(queryByText('Fim da release')).toBeInTheDocument();
  });

  it('renders the skeleton when loading', () => {
    (useRequest as jest.Mock).mockReturnValueOnce({
      data: [],
      isLoading: true
    });

    const { queryByText } = render(<ReleasesList />);
    expect(queryByText('Nome')).not.toBeInTheDocument();
  });

  it('render empty array when data is undefined', () => {
    (useRequest as jest.Mock).mockReturnValueOnce({
      data: undefined,
      isLoading: false
    });

    const { queryByText } = render(<ReleasesList />);
    expect(queryByText('Nome')).toBeInTheDocument();
  });

  it('call pushToReleasesPath when click on button', () => {
    const { getByRole } = render(<ReleasesList />);
    const button = getByRole('button', { name: 'VER MAIS...' });
    fireEvent.click(button);
    expect(useRouter().push).toBeCalledWith('/products/1-1-MeasureSoftGram/releases')
  });
});
