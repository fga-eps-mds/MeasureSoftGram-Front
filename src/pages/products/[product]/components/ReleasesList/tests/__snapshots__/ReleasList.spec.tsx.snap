// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReleasesList /> renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-m6nhw9"
  >
    <div
      class="styles__Wrapper-sc-f624tf-0 dOldAl"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-3ta6hh-MuiTypography-root"
        >
          Última Release criada
        </h5>
        <div
          class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
        >
          <table
            class="MuiTable-root css-rqglhn-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Nome
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Início da release
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Fim da release
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
        <div
          class="MuiBox-root css-1aobbod"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            VER MAIS...
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
